Core/source/core/pgp-key.ts

```diff
@@ -349,32 +364,53 @@ export class PgpKey {
    * Get latest self-signature date, in utc millis.
    * This is used to figure out how recently was key updated, and if one key is newer than other.
    */
-  public static lastSig = async (key: OpenPGP.key.Key): Promise<number> => {
-    await key.getExpirationTime(); // will force all sigs to be verified
-    const allSignatures: OpenPGP.packet.Signature[] = [];
+  public static lastSig = async (key: Key): Promise<number> => {
+    // "await key.getExpirationTime()" no longer works, need some alternate solution
+    // discussion is in progress: https://github.com/openpgpjs/openpgpjs/discussions/1491
+    const allSignatures: SignaturePacket[] = [];
     for (const user of key.users) {
-      allSignatures.push(...user.selfCertifications);
+      const data = { userID: user.userID, userAttribute: user.userAttribute, key: key };
+      for (const selfCert of user.selfCertifications) {
+        try {
+          // todo - add proper type in openpgp.d.ts
+          await selfCert.verify(key.keyPacket, enums.signature.certGeneric, data as any); // quick hack
+          allSignatures.push(selfCert);
+        } catch (e) {
+          console.log(`PgpKey.lastSig: Skipping self-certification signature because it is invalid: ${String(e)}`);
+        }
+      }
     }
-    for (const subKey of key.subKeys) {
-      allSignatures.push(...subKey.bindingSignatures);
+    for (const subKey of key.subkeys) {
+      try {
+        const latestValidKeyBindingSig = await subKey.verify();
+        allSignatures.push(latestValidKeyBindingSig);
+      } catch (e) {
+        console.log(`PgpKey.lastSig: Skipping subkey ${subKey.getKeyID().toHex()} ` +
+          `because there is no valid binding signature: ${String(e)}`);
+      }
     }
-    allSignatures.sort((a, b) => b.created.getTime() - a.created.getTime());
-    const newestSig = allSignatures.find(sig => sig.verified === true);
-    if (newestSig) {
-      return newestSig.created.getTime();
+    if (allSignatures.length > 0) {
+      allSignatures.sort((a, b) => {
+        return (b.created ? b.created.getTime() : 0) - (a.created ? a.created.getTime() : 0);
+      });
+      const newestSig = allSignatures[0];
+      return newestSig.created ? newestSig.created.getTime() : 0;
     }
     throw new Error('No valid signature found in key');
   }
 
-  public static revoke = async (key: OpenPGP.key.Key): Promise<string | undefined> => {
-    if (! await key.isRevoked()) {
-      key = await key.revoke({});
+  public static revoke = async (key: Key): Promise<string | undefined> => {
+    // todo - quick hack to pass tests
+    if (!key.isRevoked(undefined as any as SignaturePacket)) {
```

Comment:
@IvanPizhenko I added this while debugging the imports - it's wrong as is - please have a look

Reply:
Because you removed copying modified `openpgp.d.ts` in the build script into node_modules, it doesn't pick up rest of my changes. These changes are submitted as PRs to openpgp.js and some are even merged. However what you've done here also works, and it must stay here until openpgp.js releases new version with my edits to `openpgp.d.ts`.

Reply:
https://github.com/FlowCrypt/flowcrypt-ios/issues/1377